### PR TITLE
sysvar, config: introduce tidb_connection_concurrency_limit as an instance scope sysvar.

### LINKED ISSUE #34583: Make token-limit configurable as an INSTANCE scoped variable
## Enhancement

This is forked from https://github.com/pingcap/tidb/issues/33769

For `token-limit` there is a use-case to make it online configurable, but it's not a great fit for `GLOBAL`. This enhancement proposes that we add an instance scoped variable for it.

For the name, I suggest `tidb_max_connection_concurrency` or `tidb_connection_concurrency_limit`.

This is different from the issues in https://github.com/pingcap/tidb/issues/33769 because it is not a breaking change. It can still be configured in the config file as `token-limit` or via the instance scoped variable that we are introducing.

### PR DESCRIPTION
**Pull request by CbcWestwolf**
### What problem does this PR solve?

Issue Number: close #34583 

Problem Summary:

For `token-limit` there is a use-case to make it online configurable. And the name of `token-limit` is too tied to an implementation, not what the feature does.

### What is changed and how it works?

Introducing `tidb_connection_concurrency_limit` as a sysvar and a config item in `[instance]` section.

### Check List

Tests 

- [x] Unit test (added in previous UT)
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [x] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by morgo** [comment]
@CbcWestwolf Please fix linter issues:
```

[2022-05-18T15:07:18.743Z] linting

[2022-05-18T15:07:18.743Z]   ✘  https://revive.run/r#exported  exported method Server.UpdateTokenLimiter should have comment or be unexported  

[2022-05-18T15:07:18.743Z]   server/server.go:164:1

[2022-05-18T15:07:18.743Z] 

[2022-05-18T15:07:18.743Z]   ✘  https://revive.run/r#exported  exported method TokenLimiter.Resize should have comment or be unexported  

[2022-05-18T15:07:18.743Z]   server/tokenlimiter.go:46:1

[2022-05-18T15:07:18.743Z] 

[2022-05-18T15:07:18.743Z] ✘ 2 problems (2 errors, 0 warnings)

[2022-05-18T15:07:18.743Z] 

[2022-05-18T15:07:18.743Z] Errors:

[2022-05-18T15:07:18.743Z]   2  exported  
```

File: server/conn.go [hunk 1]
```diff
@@ -1324,6 +1324,7 @@ func (cc *clientConn) dispatch(ctx context.Context, data []byte) error {
 		}
 
 		cc.server.releaseToken(token)
+		cc.server.UpdateTokenLimiter()
```
Comment (morgo) [inline server/conn.go:1327]: I'm curious why update here and not in the SetGlobal func?

**Comment by CbcWestwolf** [comment]
> @CbcWestwolf Please fix linter issues:

Thanks, it is updated. PTAL. 
And I'm adding some tests to check the `UpdateTokenLimiter`. I'm afraid that this PR introduce a great performance regression due to the `mutex` in `TokenLimiter`, do you think so?

**Comment by morgo** [comment]
> Thanks, it is updated. PTAL.
> And I'm adding some tests to check the `UpdateTokenLimiter`. I'm afraid that this PR introduce a great performance regression due to the `mutex` in `TokenLimiter`, do you think so?

I was actually wondering the same thing. I think we might need to test it.

The other option is to refactor to use an atomic for the limit, and not use a resizer. It probably does change quite a bit of code though, so I'm not sure if its better.

**Review by morgo — APPROVED** [review]
LGTM, please fix conflicts

**Pull request closed without merging**